crates/ra_assists/src/handlers/extract_struct_from_enum_variant.rs

```diff
@@ -0,0 +1,341 @@
+    SourceFile, SyntaxNode, TextRange, TextSize,
+};
+use stdx::format_to;
+
+use crate::{
+    assist_context::{AssistBuilder, AssistDirector},
+    utils::insert_use_statement,
+    AssistContext, AssistId, Assists,
+};
+use ast::{ArgListOwner, VisibilityOwner};
+use hir::{EnumVariant, Module, ModuleDef};
+use ra_db::FileId;
+use ra_fmt::leading_indent;
+use rustc_hash::FxHashSet;
+
+// Assist extract_struct_from_enum
+//
+// Extracts a from struct from enum variant
+//
+// ```
+// enum A { <|>One(u32, u32) }
+// ```
+// ->
+// ```
+// struct One(pub u32, pub u32);
+//
+// enum A { One(One) }"
+// ```
+pub(crate) fn extract_struct_from_enum(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let variant = ctx.find_node_at_offset::<ast::EnumVariant>()?;
+    let field_list = match variant.kind() {
+        ast::StructKind::Tuple(field_list) => field_list,
+        _ => return None,
+    };
+    let variant_name = variant.name()?.to_string();
+    let enum_ast = variant.parent_enum();
+    let enum_name = enum_ast.name().unwrap().to_string();
+    let visibility = enum_ast.visibility();
+    let variant_hir = ctx.sema.to_def(&variant)?;
+
+    if existing_struct_def(ctx.db, &variant_name, &variant_hir) {
+        return None;
+    }
+
+    let target = variant.syntax().text_range();
+    return acc.add_in_multiple_files(
+        AssistId("extract_struct_from_enum_variant"),
+        "Extract struct from enum variant",
+        target,
+        |edit| {
+            let definition = Definition::ModuleDef(ModuleDef::EnumVariant(variant_hir));
+            let res = definition.find_usages(&ctx.db, None);
+            let module_def = mod_def_for_target_module(ctx, &enum_name);
+            let start_offset = variant.parent_enum().syntax().text_range().start();
+            let mut seen_files_map: FxHashSet<Module> = FxHashSet::default();
+            seen_files_map.insert(module_def.module(ctx.db).unwrap());
+            for reference in res {
+                let source_file = ctx.sema.parse(reference.file_range.file_id);
+                update_reference(
+                    ctx,
+                    edit,
+                    reference,
+                    &source_file,
+                    &module_def,
+                    &mut seen_files_map,
+                );
+            }
+            extract_struct_def(
+                edit,
+                enum_ast.syntax(),
+                &variant_name,
+                &field_list.to_string(),
+                start_offset,
+                ctx.frange.file_id,
+                &visibility,
+            );
+            let list_range = field_list.syntax().text_range();
+            update_variant(edit, &variant_name, ctx.frange.file_id, list_range);
+        },
+    );
+}
+
+fn existing_struct_def(db: &RootDatabase, variant_name: &str, variant: &EnumVariant) -> bool {
+    let module_defs = variant.parent_enum(db).module(db).scope(db, None);
+    for (name, _) in module_defs {
```

Comment:
You can use any for this one I think https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.any